nbconvert/filters/markdown_mistune.py

```diff
@@ -21,15 +21,58 @@
 from nbconvert.filters.strings import add_anchor
 
 
+class MathBlockGrammar(mistune.BlockGrammar):
+    """This defines a single regex comprised of the different patterns that 
+    identify math content spanning multiple lines. These are used by the 
+    MathBlockLexer.
+    """
+    multi_math_str = "|".join([r"(^\$\$.*?\$\$)",
+                               r"(^\\\\\[.*?\\\\\])",
+                               r"(^\\begin\{([a-z]*\*?)\}(.*?)\\end\{\4\})"])
+    multiline_math = re.compile(multi_math_str, re.DOTALL)
+
+
 class MathInlineGrammar(mistune.InlineGrammar):
+    """This defines different ways of declaring math objects that should be 
+    passed through to mathjax unaffected. These are used by the MathInlineLexer.
+    """
     inline_math = re.compile(r"^\$(.+?)\$|^\\\\\((.+?)\\\\\)", re.DOTALL)
     block_math = re.compile(r"^\$\$(.*?)\$\$|^\\\\\[(.*?)\\\\\]", re.DOTALL)
     latex_environment = re.compile(r"^\\begin\{([a-z]*\*?)\}(.*?)\\end\{\1\}",
                                    re.DOTALL)
     text = re.compile(r'^[\s\S]+?(?=[\\<!\[_*`~$]|https?://| {2,}\n|$)')
 
 
+class MathBlockLexer(mistune.BlockLexer):
+    """ This acts as a pass-through to the MathInlineLexer. It is needed in 
+    order to avoid other block level rules splitting math sections apart. 
+    """
+    
+    default_rules = (['multiline_math']
+                     + mistune.BlockLexer.default_rules)
+
+    def __init__(self, rules=None, **kwargs):
+        if rules is None:
+            rules = MathBlockGrammar()
+        super(MathBlockLexer, self).__init__(rules, **kwargs)
+
+    def parse_multiline_math(self, m):
+        """Add token to pass through mutiline math."""
+        self.tokens.append({
+            "type": "multiline_math",
+            "text": m.group(1) or m.group(2) or m.group(3)
```

Comment:
Since the three groups are each the whole of their alternative, this is equivalent to `m.group(0)`, right?

Reply:
Nice! Much cleaner, grazie.